### PR TITLE
perf: reduce DuckDB writer contention on the write side (#655)

### PR DESCRIPTION
**Pull request by norrietaylor**
## Root cause (issue #655)

Under a concurrent feed poll (~79 sources) plus gh-sync, the single DuckDB writer is saturated: a Logfire trace showed reads (get/search/find_similar) blocking ~48s on the writer lock. Two compounding write-side problems:

1. **Unbounded poll fan-out.** `FeedPoller.poll` fanned out over all sources with an unbounded `asyncio.gather`, saturating the thread pool and the writer lock for tens of seconds per cycle and self-inflicting upstream 429/502s.
2. **No DuckDB resource caps.** No `memory_limit` / `threads` PRAGMA was ever set, so DuckDB sized its buffer manager off total host RAM and spawned one worker per core — both unbounded, with prior OOM history (the limit was bumped 512MB → 1024 → 2048 in earlier incidents).
3. **WAL checkpoint starvation.** With overlapping write transactions, the per-write plain `CHECKPOINT` can never acquire the lock, the WAL grows unbounded, and every read replays an ever-larger WAL → latency degrades monotonically ("slower over time" symptom).

A separate PR ([#670](https://github.com/norrietaylor/distillery/pull/670)) moved the *reads* off the writer lock. This PR is the **write side**: make the writer cheaper / shorter-held and bound the poll's resource use so the writer isn't saturated for ~48s in the first place.

### A. Bound feed-poller concurrency (`perf(feeds)`)
- Added an `asyncio.Semaphore` (default **5**, constructor-overridable via `max_concurrent_sources`, clamped to ≥ 1) around `_poll_source`, so at most N sources are fetched + scored + stored concurrently.
- Behaviour preserved: every source is still polled; only peak concurrency is bounded. This cuts CPU / thread-pool saturation, self-inflicted upstream 429/502s, and writer contention.

### B. DuckDB resource PRAGMAs at connection init (`perf(store)`)
- Set explicit `memory_limit` (**~1GB**) and `threads` (**=2**) PRAGMAs in `_open_connection`, with safe defaults that are overridable via new constructor kwargs (`memory_limit_mb`, `threads`). `None` / non-positive leaves DuckDB's default in place. Best-effort: a PRAGMA failure is logged, not fatal.

### C. Idle-window checkpoint cadence (`perf(store)`) — implemented, safely
This is the riskiest part, so it was implemented conservatively. I read the existing checkpoint logic and history first: recent commits **deliberately dropped** a "FORCE CHECKPOINT escalation that could deadlock the writer" (issue #648). This change **does not** reintroduce that.

- Added an **opt-in background idle-checkpoint loop**, started by the server lifespan after `store.initialize()` and stopped in `close()`.
- It issues a **plain `CHECKPOINT` only during quiet windows**: it pre-checks `_conn_lock.locked()` and **skips** the sweep whenever a writer is active, so it never blocks or queues behind a writer.
- It holds `_conn_lock` only for the (non-blocking) plain `CHECKPOINT`, preserving the single-connection-one-thread invariant (issue #416). It **never** escalates to `FORCE CHECKPOINT`. No-op for in-memory DBs.
- Net effect: the WAL gets flushed during quiet windows so it can't grow unbounded, without any deadlock or writer-blocking risk.

## Tests
- `tests/test_poller.py::TestFeedPollerConcurrencyBound` — tracks in-flight fetch count via a patched adapter and asserts peak concurrency never exceeds the limit (while all sources are still polled); plus the clamp-to-1 guard.
- `tests/test_duckdb_store.py::TestResourcePragmas` — asserts `current_setting('memory_limit')` / `current_setting('threads')` reflect the applied PRAGMAs (default, overridden, and disabled paths).
- `tests/test_duckdb_store.py::TestIdleCheckpoint` — asserts the idle sweep flushes the WAL, **skips when the writer lock is held** (never blocks), start/stop lifecycle + idempotency, and the in-memory no-op.

## Results
- `ruff check src/ tests/` — clean
- `ruff format` — applied to changed files
- `mypy --strict src/distillery/` — clean (73 source files)
- `pytest -q` — **3143 passed, 98 skipped** (full suite)
- Diff stat: 5 files changed, ~419 insertions, ~16 deletions. `uv.lock` untouched.

Refs #655

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a configurable limit on how many feed sources are processed concurrently per polling cycle.
  * Introduced optional idle-time database WAL checkpointing to maintain performance during quiet periods.

* **Bug Fixes**
  * Reduced database WAL growth under concurrent writers and improved overall stability during heavy polling.
  * Improved initialization and shutdown behavior so database maintenance starts/stops reliably.

* **Tests**
  * Added coverage for concurrency bounding and idle checkpoint behavior, including edge cases (e.g., in-memory databases).

### CODE REVIEW

File: src/distillery/store/duckdb.py [hunk 1]
```diff
@@ -1185,6 +1248,91 @@ async def close(self) -> None:
...
+            return
+        async with lock:
+            conn = self._conn
+            if conn is None:
+                return
+            try:
+                await asyncio.to_thread(conn.execute, "CHECKPOINT")
+                self._checkpoint_failures = 0
+                logger.debug("Idle checkpoint flushed WAL")
+            except duckdb.Error as exc:
```
[the remark this answers is not in the document]
Comment (norrietaylor) [reply src/distillery/store/duckdb.py:1340]: Fixed in 1b77026 — the idle-checkpoint loop now runs each sweep as a shielded child, so a cancel from stop_idle_checkpoint()/close() waits for the in-flight CHECKPOINT worker to finish before the task ends; no orphan thread races the connection teardown (same fix as #670's flusher).

File: tests/test_duckdb_store.py [hunk 2]
```diff
@@ -2315,6 +2310,162 @@ async def batch_writer(n: int) -> None:
...
+            with patch.object(store, "_checkpoint_after_write", lambda *_a, **_k: None):
+                for i in range(5):
+                    await store.store(make_entry(content=f"idle-{i}"))
+            wal_path = Path(db_path + ".wal")
+            wal_before = wal_path.stat().st_size if wal_path.exists() else 0
+
+            await store._idle_checkpoint_once()  # noqa: SLF001
+
+            wal_after = wal_path.stat().st_size if wal_path.exists() else 0
+            assert wal_after < wal_before or wal_after == 0
```
[the remark this answers is not in the document]
Comment (norrietaylor) [reply tests/test_duckdb_store.py]: Fixed in 1b77026 — replaced the env-flaky WAL-size assertion (which passed vacuously when the WAL was 0) with a caplog assertion that the sweep actually executed the CHECKPOINT success path.
